File: app/src/main/java/com/haoyu/app/fragment/OfficeViewerFragment.java
package com.haoyu.app.fragment;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebView;
import android.widget.FrameLayout;

import com.haoyu.app.lego.student.R;
import com.haoyu.app.view.ProgressWebView;

/**
 * 创建日期：2017/11/16.
 * 描述:office文档查看
 * 作者:xiaoma
 */

public class OfficeViewerFragment extends Fragment {

    private FrameLayout fl_content;
    private FrameLayout fl_failure;
    private ProgressWebView webView;
    private Context context;
    private String url;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.context = context;
        Bundle bundle = getArguments();
        String url = bundle.getString("url");
        this.url = "https://view.officeapps.live.com/op/view.aspx?src=" + url;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_webviewer, container, false);
        initView(view);
        return view;
    }

    private void initView(View view) {
        fl_content = view.findViewById(R.id.fl_content);
        fl_failure = view.findViewById(R.id.fl_failure);
        configWebview(url);
    }

    private void configWebview(String url) {
        webView = new ProgressWebView(context);
        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.MATCH_PARENT);
        webView.setLayoutParams(params);
        fl_content.addView(webView);
        webView.setOnReceivedListener(new ProgressWebView.OnReceivedListener() {
            @Override
            public void onReceivedTitle(WebView view, String title) {

            }

            @Override
            public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
                fl_content.setVisibility(View.GONE);
                fl_failure.setVisibility(View.VISIBLE);
            }

            @Override
            public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
                fl_content.setVisibility(View.GONE);
                fl_failure.setVisibility(View.VISIBLE);
            }
        });
        webView.setLayerType(View.LAYER_TYPE_SOFTWARE, null);
        webView.loadUrl(url);
        webView.setOnKeyListener(new View.OnKeyListener() {
            @Override
            public boolean onKey(View v, int keyCode, KeyEvent event) {
                if (event.getAction() == KeyEvent.ACTION_DOWN) {
                    if (keyCode == KeyEvent.KEYCODE_BACK && webView.canGoBack()) {  //表示按返回键
                        webView.goBack();   //后退
                        return true;    //已处理
                    }
                }
                return false;
            }
        });
        fl_failure.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                fl_content.setVisibility(View.VISIBLE);
                fl_failure.setVisibility(View.GONE);
                webView.reload();
            }
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        if (webView != null) {
            webView.onResume();
            webView.resumeTimers();
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        if (webView != null) {
            webView.onPause();
            webView.pauseTimers(); //小心这个！！！暂停整个 WebView 所有布局、解析、JS。
        }
    }

    @Override
    public void onDestroyView() {
        fl_content.removeAllViews();
        if (webView != null) {
            webView.stopLoading();
            webView.removeAllViews();
            webView.destroy();
            webView = null;
        }
        super.onDestroyView();
    }
}

File: app/src/main/java/com/haoyu/app/fragment/OfficeViewerFragment.kt
package com.haoyu.app.fragment

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebChromeClient
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.FrameLayout
import android.widget.ProgressBar
import com.haoyu.app.lego.student.R
import com.haoyu.app.view.LoadFailView

@Suppress("DEPRECATION")
/**
 * 创建日期：2018/1/18.
 * 描述:office文档查看
 * 作者:xiaoma
 */
class OfficeViewerFragment : Fragment() {
    private lateinit var flContent: FrameLayout
    private lateinit var webView: WebView
    private lateinit var progressBar: ProgressBar
    private lateinit var loadFailView: LoadFailView
    private var url: String? = null

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        arguments?.getString("url")?.let {
            url = "https://view.officeapps.live.com/op/view.aspx?src=" + it
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_webviewer, container, false)
        initView(view)
        return view
    }

    private fun initView(view: View) {
        flContent = view.findViewById(R.id.fl_content)
        loadFailView = view.findViewById(R.id.loadFailView)
        webView = view.findViewById(R.id.webView)
        progressBar = view.findViewById(R.id.progressBar)
        loadFailView.setErrorMsg("网页加载失败，请点击重试！")
        url?.let {
            configWebview(it)
        }
    }

    private fun configWebview(url: String) {
        initWebViewSettings()
        webView.webViewClient = X5WebViewClient()
        webView.webChromeClient = X5WebChromeClient()
        webView.loadUrl(url)
        webView.setOnKeyListener(View.OnKeyListener { _, keyCode, _ ->
            if (keyCode == KeyEvent.KEYCODE_BACK && webView.canGoBack()) {
                webView.goBack()// 返回前一个页面
                return@OnKeyListener true
            }
            false
        })
        loadFailView.setOnRetryListener {
            webView.reload()
            flContent.visibility = View.VISIBLE
        }

    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun initWebViewSettings() {
        webView.settings.setSupportZoom(true)
        webView.settings.loadWithOverviewMode = true
        webView.settings.useWideViewPort = true//设置此属性，可任意比例缩放
        //调用JS方法.安卓版本大于17,加上注解 @JavascriptInterface
        webView.settings.javaScriptEnabled = true
        webView.settings.setSupportMultipleWindows(true)
        webView.settings.builtInZoomControls = true
        // 支持通过js打开新的窗口
        webView.settings.javaScriptCanOpenWindowsAutomatically = true
        webView.settings.pluginState = WebSettings.PluginState.ON
        //不显示webview缩放按钮
        webView.settings.displayZoomControls = false
        webView.settings.defaultZoom = WebSettings.ZoomDensity.FAR// 屏幕自适应网页,如果没有这个，在低分辨率的手机上显示可能会异常
        webView.settings.cacheMode = WebSettings.LOAD_CACHE_ELSE_NETWORK
        webView.settings.domStorageEnabled = true
    }

    private inner class X5WebViewClient : WebViewClient() {
        override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
            view.loadUrl(url)
            return true
        }

        override fun onReceivedError(view: WebView, errorCode: Int, description: String, failingUrl: String) {
            flContent.visibility = View.GONE
            loadFailView.visibility = View.VISIBLE
        }
    }

    private inner class X5WebChromeClient : WebChromeClient() {

        override fun onProgressChanged(view: WebView, newProgress: Int) {
            if (loadFailView.visibility != View.GONE) loadFailView.visibility = View.GONE
            if (newProgress == 100) {
                progressBar.visibility = View.GONE
            } else {
                if (progressBar.visibility != View.VISIBLE) progressBar.visibility = View.VISIBLE
                progressBar.progress = newProgress
            }
            super.onProgressChanged(view, newProgress)
        }
    }

    override fun onResume() {
        super.onResume()
        webView.onResume()
        webView.resumeTimers()
    }

    override fun onPause() {
        super.onPause()
        webView.onPause()
        webView.pauseTimers()
    }

    override fun onDestroy() {
        super.onDestroy()
        webView.stopLoading()
        webView.removeAllViews()
        webView.destroy()
    }
}